### ISSUE TITLE
adjust header section

### ISSUE THREAD

**Issue by Alexander-Barth** [Alexander-Barth/NetCDF3#5]
As a follow-up of: https://github.com/Alexander-Barth/NetCDF3.jl/pull/4

If I understand correctly, in scipy the data is written to disk only when the file is closed:
https://github.com/scipy/scipy/blob/v1.9.0/scipy/io/_netcdf.py

At this stage one knows already the correct header size, but it also means that one needs to keep all variables in memory, which would be a problem for big netcdf files.

I guess we would need a function to shift all the data (after the header) toward the end of the file by chunks if the header section is too large. This sounds a bit inefficient. Would there be another solution?

**Comment by Alexander-Barth** [comment]
CC @visr

**Comment by visr** [comment]
Only thing I can think of is to have a temporary file that has no header but starts directly at the data. Then at writing time, we open the definitive file, write the header from memory, and streaming append the temporary file.

We could optionally have this data IO be an `IOBuffer` if keeping it in memory is not an issue. Or perhaps we could come up with something fancy based on https://github.com/JuliaData/MemPool.jl.

**Comment by Alexander-Barth** [comment]
I have now implemented the in-place data shifting for large header sections. I guess that we need this functionality anyways later (when having multiple unlimited variable). 
There is now also an optional `header_size_hint` parameter.
When you run with `ENV["JULIA_DEBUG"] = "NetCDF3"`, you will get the optimal value of this parameter when the file is closed.

**Comment by visr** [comment]
Thanks, that looks very useful. Perhaps we could also simplify this:

```julia
memio = IOBuffer()
offset0 = nc.header_size_hint

Toffset = (nc.version == 1 ? Int32 : Int64)
Tsize = (nc.version < 5 ? Int32 : Int64)

start = try_write_header(memio,nc.recs,nc.dim,nc.attrib,nc.vars,Toffset,Tsize,offset0)
header = take!(memio)
```

To something like `header = write_header(nc)`, such that it is easy to write the header or check the size programmatically.

**Comment by Alexander-Barth** [comment]
Good idea, I implemented `nc_header_size` to get the optimal value of `header_size_hint`.
